src/join/core.jl

```diff
@@ -587,98 +595,108 @@ end
                             seen_rows::AbstractVector{Bool},
-                            ::Val{right_shorter}) where {T, right_shorter}
-    if right_shorter
-        @assert length(left) == length(seen_rows)
-    else
-        @assert length(right) == length(seen_rows)
-    end
-
-    dict = Dict{T, Int}()
-
+                            right_shorter::Bool) where {T}
     right_len = length(right)
-    sizehint!(dict, 2 * min(right_len, typemax(Int) >> 2))
-
     right isa OnCol && _prehash(right)
     left isa OnCol && _prehash(left)
 
-    for (idx_r, val_r) in enumerate(right)
-        haskey(dict, val_r) && return _semijoin_dup(left, right, dict, idx_r,
-                                                    seen_rows, Val(right_shorter))
-        dict[val_r] = idx_r
-    end
-
-    @inbounds for (idx_l, val_l) in enumerate(left)
-        # we know that dict contains only positive values
-        idx_r = get(dict, val_l, -1)
-        if idx_r != -1
-            if right_shorter
-                seen_rows[idx_l] = true
-            else
+    if right_shorter
+        @assert length(left) == length(seen_rows)
+        set = Set{T}()
+        sizehint!(set, 2 * min(right_len, typemax(Int) >> 2))
+        for val_r in right
+            push!(set, val_r)
+        end
+        @inbounds for (idx_l, val_l) in enumerate(left)
+            seen_rows[idx_l] = val_l in set
+        end
+    else
+        @assert length(right) == length(seen_rows)
+        dict = Dict{T, Int}()
+        sizehint!(dict, 2 * min(right_len, typemax(Int) >> 2))
+        for (idx_r, val_r) in enumerate(right)
+            haskey(dict, val_r) && return _semijoin_dup(left, right, dict, idx_r,
+                                                        seen_rows)
+            dict[val_r] = idx_r
+        end
+        @inbounds for (idx_l, val_l) in enumerate(left)
+            # we know that dict contains only positive values
+            idx_r = get(dict, val_l, -1)
+            if idx_r != -1
                 seen_rows[idx_r] = true
             end
         end
     end
+
     return seen_rows
 end
 
 function _semijoin_unsorted_int(left::AbstractVector{<:Union{Integer, Missing}},
                                 right::AbstractVector{<:Union{Integer, Missing}},
                                 seen_rows::AbstractVector{Bool},
-                                ::Val{right_shorter}) where {right_shorter}
+                                right_shorter::Bool)
     minv, maxv = extrema_missing(right)
 
     val_range = big(maxv) - big(minv)
     if val_range > typemax(Int) - 3 || val_range ÷ 2 > max(64, length(right)) ||
        minv < typemin(Int) + 2 || maxv > typemax(Int) - 3
-       return _semijoin_unsorted(left, right, seen_rows, Val(right_shorter))
+       return _semijoin_unsorted(left, right, seen_rows, right_shorter)
     end
 
     offset = 1 - Int(minv) # we are now sure it does not overflow
     len = Int(maxv) - Int(minv) + 2
     dict = zeros(Int, len)
```

Comment:
Calling this "dict" is a bit confusing given that it's a vector. :-)

Reply:
OK - I have changed it to `group_map`. Is this better?

Reply:
Yes!